### PR TITLE
(feat) enable actions to enhance typings on applied element

### PR DESCRIPTION
**Pull request by dummdidumm**
#1243
#431

This assumes we also enhance the typing of the action type in Svelte core so they match. Assuming we have this:

```ts
export interface ActionReturn<Parameter = any, Attributes extends Record<string, any> = Record<never, any>> {
	update?: (parameter: Parameter) => void;
	destroy?: () => void;
	$$attributes?: Attributes;
}
```

You could define an action like this:

```ts
export function action(node: HTMLElement): ActionReturn<never, { foo: boolean; 'on:bar': EventHandler<Event> }> {
   // ..
}
```

and it would enhance the DOM element with the new required attribute `foo` and the optional `on:bar` event.

This requires newer TS types so we can only merge that right before the new transformation (also see https://github.com/sveltejs/language-tools/issues/1552)

Question (cc @jasonlyu123 @ignatiusmb @ivanhofer ):
Is the assumed `ActionReturn` good like this? Should we put both event handler and attributes into one parameter (people would have to type `onbar` or `on:bar` then (undecided yet what we do here, see other issues)? If not, are the typings for Events good? Should it be `bar: Event` instead and we add the `EventHandler` typing ourselves (but how to get currentTarget of the correct DOM element into that then?)

### CODE REVIEW

File: packages/svelte2tsx/src/htmlxtojsx_v2/nodes/Element.ts [hunk 1]
```diff
@@ -244,6 +238,96 @@ export class Element {
...
+        const createElement = `${this.typingsNamespace}.createElement`;
+        const addActions = () => {
+            if (this.actionIdentifiers.length) {
+                return `, __sveltets_2_union(${this.actionIdentifiers.join(',')})`;
+            } else {
+                return '';
+            }
+        };
+
+        switch (this.node.name) {
```
Comment (ivanhofer) [inline packages/svelte2tsx/src/htmlxtojsx_v2/nodes/Element.ts:251]: From what I can see the `if-else` statements are quite repetitive and error-prone on changes.
I would write it like this:

```ts
const getCreateElementStatement = (name: string): string => {
   // switch-case statement with just the else branches
   // as a single string and not as an array (from what I see it is always 1 element)
   // and without the leading '{'
}

const createElementStatement = getCreateElementStatement(this.node.name)
const variable = this.referencedName ? `const ${this._name} = ` : ''

return [`{ ${variable}${createElementStatement}`]
```

File: packages/svelte2tsx/test/svelte2tsx/samples/$store-as-directive/expectedv2.ts [hunk 2]
```diff
@@ -9,8 +9,8 @@
 ;
 async () => {
 
- { svelteHTML.createElement("div", {      });__sveltets_2_ensureTransition($transitionStore(svelteHTML.mapElementTag('div'),({ y: 100 })));__sveltets_2_ensureAction($actionStore(svelteHTML.mapElementTag('div')));__sveltets_2_ensureTransition($inStore(svelteHTML.mapElementTag('div')));__sveltets_2_ensureTransition($outStore(svelteHTML.mapElementTag('div')));__sveltets_2_ensureAnimation($animateStore(svelteHTML.mapElementTag('div'),__sveltets_2_AnimationMove));
- }};
+ {const $$action_0 = __sveltets_2_ensureAction($actionStore(svelteHTML.mapElementTag('div')));{ svelteHTML.createElement("div", __sveltets_2_union($$action_0), {      });__sveltets_2_ensureTransition($transitionStore(svelteHTML.mapElementTag('div'),({ y: 100 })));__sveltets_2_ensureTransition($inStore(svelteHTML.mapElementTag('div')));__sveltets_2_ensureTransition($outStore(svelteHTML.mapElementTag('div')));__sveltets_2_ensureAnimation($animateStore(svelteHTML.mapElementTag('div'),__sveltets_2_AnimationMove));
```
Comment (ivanhofer) [inline packages/svelte2tsx/test/svelte2tsx/samples/$store-as-directive/expectedv2.ts:12]: I can only see tests for `$$action_0`. I think a test with multiple actions applied to a dom element would be good.

What happens if two actions expect the same `attribute` with a different type?

**Review by ivanhofer — COMMENTED** [review]
The idea behind this PR is great!

I didn't dug deep into how the language-tools work so I can't give the best review. I tried to just look at the new additions and the big picture make sense to me.

You mentioned you want to enhance the `ActionReturn` type from `'svelte/action'`. Please also consider enhancing the `Action` type with the same additions.

Because I don't fully know how the language tools work, I'm wondering why a lot of (or all) types get a `__svelte...` name. When I set something a wrong param on an element I sometimes get an error inside VS Code like "x is not assignable to __svelte_xyz" where it actually could say "x is not assignable to Xyz". Is the prefix really needed?

**Comment by dummdidumm** [comment]
Thanks for the review! I wasn't expecting a review of the code (although I greatly appreciate it), more of the general idea and thoughts on the questions I mentioned - if you prefer attributes and events to be split or not etc.

File: packages/svelte2tsx/test/svelte2tsx/samples/$store-as-directive/expectedv2.ts (hunk 2 above)
Comment (dummdidumm) [reply packages/svelte2tsx/test/svelte2tsx/samples/$store-as-directive/expectedv2.ts:12]: A test with multiple actions is a good idea, yes. If actions conflict then, well, nothing we can do about that 😄 If they types conflict that likely means these actions aren't compatible at runtime, too.

File: packages/svelte2tsx/src/htmlxtojsx_v2/nodes/Element.ts (hunk 1 above)
Comment (dummdidumm) [reply packages/svelte2tsx/src/htmlxtojsx_v2/nodes/Element.ts:251]: Yeah I was thinking about cleaning this up, too, this is a good outline of the solution

**Comment by ivanhofer** [comment]
> Thanks for the review! I wasn't expecting a review of the code (although I greatly appreciate it), more of the general idea and thoughts on the questions I mentioned - if you prefer attributes and events to be split or not etc.

Sorry, I completely forgot to answer the questions from the description 😅

I prefer to have `Attributes` and `Events` split. If not possible/wanted, I would prefer `on:event` instead of the `jsx`-like `onEvent`, since it is 1:1 what you write on the element when listening to the event. And `onEvent` could also confilct with a prop that is called the same.

The `ActionReturn` interface should probably restrict `Attributes` and `Events` to be `Records`:
```ts
export interface ActionReturn<Parameter = any, Attributes extends Record<string, any> = {}, Events extends Record<string, any> = {}> {
	...
}
```

Defining `Events` should be handled the same way like in `createEventDispatcher`.

> About the prefixes: They exist so it's easier to filter them out of completions, and because this minimizes the probability that they clash with other definitions - they are global/ambient so everyone could use the types by accident.

By accident or as a workaround. And the naming can change. I today realized that [this code](https://github.com/ivanhofer/sveltekit-typescript-showcase/blob/main/src/01-props/09-svelte-component/Component.svelte#L12) does not work anymore because the types have changed in a recent version. (But I can soon switch to the new types from https://github.com/sveltejs/svelte/pull/6770 to get rid of those types)

**Comment by dummdidumm** [comment]
> By accident or as a workaround. And the naming can change. I today realized that [this code](https://github.com/ivanhofer/sveltekit-typescript-showcase/blob/main/src/01-props/09-svelte-component/Component.svelte#L12) does not work anymore because the types have changed in a recent version. (But I can soon switch to the new types from https://github.com/sveltejs/svelte/pull/6770 to get rid of those types)

Yeah that's an unfortunate outcome of these types being global. People come to the idea that they can use that but that wasn't the intention, and then they get broken code after bumps. Maybe as part of the new transformation this all should live in a namespace which we hide from autocompletion.

**Comment by ignatiusmb** [comment]
> Is the assumed `ActionReturn` good like this? Should we put both event handler and attributes into one parameter (people would have to type `onbar` or `on:bar` then (undecided yet what we do here, see other issues)?

Ideally they're specified in one place, I would like to avoid the need to skip a parameter as much as possible. I think we should be able to make it so that properties that starts with `on:` would be treated as events, otherwise an attribute by default.

> are the typings for Events good? Should it be `bar: Event` instead and we add the `EventHandler` typing ourselves (but how to get currentTarget of the correct DOM element into that then?)

I'd say it's probably better to be as explicit as possible, especially when working with TS. More so if it makes the implementation on the language-tools side more complicated.

**Comment by dummdidumm** [comment]
Since the event HTML typings in https://github.com/sveltejs/svelte/pull/7649 are now of the form `on:X` I agree with @ignatiusmb that we should make this one generic where you type both names and events. Will open a PR over at Svelte core shortly.

**Pull request merged**